packages/cozy-stack-client/src/OAuthClient.js

```diff
@@ -0,0 +1,358 @@
+import CozyStackClient from './CozyStackClient'
+import AccessToken from './AccessToken'
+
+const defaultoauthOptions = {
+  clientID: '',
+  clientName: '',
+  clientKind: '',
+  clientSecret: '',
+  clientURI: '',
+  registrationAccessToken: '',
+  redirectURI: '',
+  softwareID: '',
+  softwareVersion: '',
+  logoURI: '',
+  policyURI: '',
+  notificationPlatform: '',
+  notificationDeviceToken: ''
+}
+
+export default class OAuthClient extends CozyStackClient {
+  constructor({ oauth, scope, ...options }) {
+    super(options)
+    this.oauthOptions = { ...defaultoauthOptions, ...oauth }
+    this.scope = scope
+  }
+
+  /**
+   * Checks if the client has his registration informations from the server
+   * @returns {boolean}
+   * @private
+   */
+  isRegistered() {
+    return this.oauthOptions.clientID !== ''
+  }
+
+  /**
+   * Converts a camel-cased data set to snake case, suitable for sending to the OAuth server
+   * @param   {object} data Initial data
+   * @returns {object} Formatted data
+   * @private
+   */
+  snakeCaseOAuthData(data) {
+    const mappedFields = {
+      softwareID: 'software_id',
+      softwareVersion: 'software_version',
+      clientID: 'client_id',
+      clientName: 'client_name',
+      clientKind: 'client_kind',
+      clientURI: 'client_uri',
+      logoURI: 'logo_uri',
+      policyURI: 'policy_uri',
+      notificationPlatform: 'notification_platform',
+      notificationDeviceToken: 'notification_device_token',
+      redirectURI: 'redirect_uris'
+    }
+
+    const result = {}
+
+    Object.keys(data).forEach(fieldName => {
+      const key = mappedFields[fieldName] || fieldName
+      const value = data[fieldName]
+      result[key] = value
+    })
+
+    // special case: turn redirect_uris into an array
+    if (
+      result['redirect_uris'] &&
+      result['redirect_uris'] instanceof Array === false
+    )
+      result['redirect_uris'] = [result['redirect_uris']]
+
+    return result
+  }
+
+  /**
+   * Converts a snake-cased data set to camel case, suitable for internal use
+   * @param   {object} data Initial data
+   * @returns {object} Formatted data
+   * @private
+   */
+  camelCaseOAuthData(data) {
+    const mappedFields = {
+      client_id: 'clientID',
+      client_name: 'clientName',
+      client_secret: 'clientSecret',
+      registration_access_token: 'registrationAccessToken',
+      software_id: 'softwareID'
+    }
+
+    const result = {}
+
+    Object.keys(data).forEach(fieldName => {
+      const key = mappedFields[fieldName] || fieldName
+      const value = data[fieldName]
+      result[key] = value
+    })
+
+    return result
+  }
```

Comment:
Yet another function that could be written with a simple and more straight forward regexp. See above.